### ISSUE TITLE
Unable to `git spr merge` when target branch requires signed commits

### ISSUE THREAD

**Issue by wade13** [ejoffe/spr#195]
When the **Require signed commits** branch protection is enabled
![image](https://user-images.githubusercontent.com/96310424/154762671-d586f680-71d6-41d9-be61-556e73ed7c93.png)

 `git spr merge` fails with a GitHub error:

```
Base branch requires signed commits. Rebase merges cannot be automatically signed by GitHub`.
```

For example:

```console
[main workflow-demo]$ git spr s
> git rev-parse --show-toplevel
> github fetch pull requests
> git branch

 ┌─ github checks pass
 │ ┌── pull request approved
 │ │ ┌─── no merge conflicts
 │ │ │ ┌──── stack check
 │ │ │ │
[✅➖✅✅] https://github.com/wade13/workflow-demo/pull/21 : s4: change 2
[✅➖✅✅] https://github.com/wade13/workflow-demo/pull/20 : s4: change 1

[main workflow-demo]$ git spr merge
> git rev-parse --show-toplevel
> github fetch pull requests
> git branch
> github update 21 - s4: change 2
1:26PM FTL ../../src/github.com/wade13/spr/github/githubclient/client.go:488 > pull request merge failed error="Base branch requires signed commits. Rebase merges cannot be automatically signed by GitHub" id=PR_kwDOG26l6M4zH6aM number=21 title="s4: change 2"
```

This is a known GitHub issue (github/hub#1241). A possible workaround is to use "MERGE" instead of "REBASE" for `GitHubInterface.MergePullRequest`, presumably based on a per-repo configuration in `.spr.yml`.

This seems to work out alright in practice, for example:
https://github.com/wade13/workflow-demo/pull/20
https://github.com/wade13/workflow-demo/pull/21

It shows in wade13/workflow-demo#20 that it was merged with wade13/workflow-demo#21, but of course the user needs to be OK with using a merge commit:

```
*   2befe0e (HEAD -> main, origin/main, origin/HEAD) Merge pull request #21 from wade13/pr/wade13/main/59325325
|\  
| * 7e72768 (origin/pr/wade13/main/59325325) s4: change 2
| * 7658a7d (origin/pr/wade13/main/0c33c9d4) s4: change 1
|/  
```

**Comment by ejoffe** [comment]
@wade13  - Sounds like a good plan. Welcome to the contributors club!

**Comment by wade13** [comment]
> @wade13 - Sounds like a good plan. Welcome to the contributors club!

Thanks :smile:. I'm planning on calling the config option `rebaseMerge`, a boolean.

**Comment by ejoffe** [comment]
If we go down this path, we should allow all the different merge methods. 
I suggest we make the configuration a string and call it `mergeMethod`, and allow all options ["merge", "squash", "rebase"], default should be "rebase". 
You can then map the configuration value to the corresponding PullRequestMergeMethod values. If an invalid configuration is used, just log a an error and exit without merging.

**Comment by wwade** [comment]
> If we go down this path, we should allow all the different merge methods. I suggest we make the configuration a string and call it `mergeMethod`, and allow all options ["merge", "squash", "rebase"], default should be "rebase". You can then map the configuration value to the corresponding PullRequestMergeMethod values. If an invalid configuration is used, just log a an error and exit without merging.

Oops, saw this after, sure, I'll make that change.


Also, I _was_ trying to use `spr`, but I wasn't able to get it going from my fork. I'm guessing this is just not a supported workflow?
```
$ git spr u
> git rev-parse --show-toplevel
> git fetch
> git rebase origin/master --autostash
> github fetch pull requests
> git branch
> git log origin/master..HEAD
> git status --porcelain --untracked-files=no
> git push --force --atomic origin ca80411c9573945b41f853a5d2164d2458cc4d30:refs/heads/pr/wade13/master/95a152eb 64f7331ae723e40f1831688f81171d5a34dd818b:refs/heads/pr/wade13/master/5e16d0db
panic: Head sha can't be blank, Base sha can't be blank, No commits between master and pr/wade13/master/95a152eb, Head ref must be a branch

goroutine 1 [running]:
github.com/ejoffe/spr/github/githubclient.check({0x878d20, 0xc000614210})
	/home/wade/src/github.com/wwade/spr/github/githubclient/client.go:554 +0xfa
github.com/ejoffe/spr/github/githubclient.(*client).CreatePullRequest(0xc000056eb0, {0x880bd0, 0xc000018158}, 0xc00051e0a0, {{0xc0005664b2, 0x8}, {0xc0005661e8, 0x28}, {0xc000566275, 0x2d}, ...}, ...)
	/home/wade/src/github.com/wwade/spr/github/githubclient/client.go:309 +0x605
github.com/ejoffe/spr/spr.(*stackediff).UpdatePullRequests(0xc00006ac00, {0x880bd0, 0xc000018158})
	/home/wade/src/github.com/wwade/spr/spr/spr.go:162 +0xa98
main.main.func3(0xc000012e80)
	/home/wade/src/github.com/wwade/spr/cmd/spr/main.go:140 +0x25
github.com/urfave/cli/v2.(*Command).Run(0xc0000a5680, 0xc00029ec40)
	/home/wade/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/command.go:163 +0x64a
github.com/urfave/cli/v2.(*App).RunContext(0xc0004dc4e0, {0x880bd0, 0xc000018158}, {0xc000012040, 0x2, 0x2})
	/home/wade/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:313 +0x81e
github.com/urfave/cli/v2.(*App).Run(...)
	/home/wade/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:224
main.main()
	/home/wade/src/github.com/wwade/spr/cmd/spr/main.go:188 +0xe79
```

**Comment by wwade** [comment]
Alright @ejoffe thanks for the help on this. It should be ready to go when you have a chance to review.

**Comment by ejoffe** [comment]
Regarding your failure, I'm not sure what the issue is. What does your repo config look like?
spr should work on your fork wwade/spr, it won't work if you try to push into ejoffe/spr

**Comment by wwade** [comment]
Sorry, I shouldn't have overloaded this thread with all this noise :smile: 

I tried it with
```
githubRepoOwner: wwade
githubRepoName: spr
githubHost: github.com
requireChecks: true
requireApproval: true
githubRemote: origin
githubBranch: master
mergeMethod: rebase
```

and then it just makes PRs against `wwade/spr`.

The config I used for the error message is just
```
githubRepoOwner: ejoffe
githubRepoName: spr
githubHost: github.com
requireChecks: true
requireApproval: true
githubRemote: origin
githubBranch: master
mergeMethod: rebase
```

`origin` is my fork.

So I tried using `githubRemote: upstream` as well:
```
githubRepoOwner: wwade
githubRepoName: spr
githubHost: github.com
requireChecks: true
requireApproval: true
githubRemote: upstream
githubBranch: master
mergeMethod: rebase
```

but that gave a different error:
```
> git push --force --atomic upstream 09a97c5760a00a009ecfd78d7dd0ec00809aaee7:refs/heads/pr/wade13/config-merge-using-merge/089df4dd b361412b6dc1371225ea0c5ca6127f849ae0bb07:refs/heads/pr/wade13/config-merge-using-merge/0bb98aa6
git error: ERROR: Permission to ejoffe/spr.git denied to wwade.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
error: exit status 128
```

**Comment by ejoffe** [comment]
@wade13 - I added you as a collaborator to ejoffe/spr, you should now be able to run spr on this repo without needed your own fork. I added wade13 since that's the one you used for pull requests.

**Comment by wwade** [comment]
Thanks. Actually, mind adding @wwade instead? That's the one I ended up using for the PRs that ended up merging.

**Comment by ejoffe** [comment]
I just noticed and switched it :)

**Comment by ejoffe** [comment]
fix released in v0.8.2
